Longname: files/directories name upto 1023 bytes

tuxoko opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->
This patch adds the ability for zfs to support file/dir name up to 1023
bytes. This number is chosen so we can support up to 255 4-byte
characters. This new feature is represented by the new feature flag
feature@longname.

A new dataset property "longname" is also introduced to toggle longname
support for each dataset individually. This property can be disabled,
even if it contains longname files. In such case, new file cannot be
created with longname but existing longname files can still be looked
up.

Note that, to my knowledge native Linux filesystems don't support name
longer than 255 bytes. So there might be programs not able to work with
longname.

Note that NFS server may needs to use exportfs_get_name to reconnect
dentries, and the buffer being passed is limit to NAME_MAX+1 (256). So
NFS may not work when longname is enabled.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

amotin commented on include/sys/zap.h:
```diff
@@ -367,9 +367,20 @@ typedef struct {
 	uint64_t za_num_integers;
 	uint64_t za_first_integer;	/* no sign extension for <8byte ints */
-	char za_name[ZAP_MAXNAMELEN];
+	uint32_t za_name_len;
```
As I understand, this represents total length of the field.  Just couple days ago looking recently on ZAP code I've found that its iterator code can not properly report binary names, since this structure includes neither length nor number of integers for the name, unlike value.  I am thinking if it would be good to fix while you are here and changing the API.

tuxoko replied:
@amotin Right, this represent the buffer len for za_name.
Regarding your issue, can you elaborate a bit. Is there any command or api affect by this?

amotin replied:
@tuxoko There may be other examples, but I particularly hit that dump_zap() in zdb is unable to handle DDT and BRT ZAPs.

tonyhutter commented:
@tuxoko sorry this PR has sat on the sidelines for so long.  Can you rebase on master and I can take a look?

tuxoko commented:
@tonyhutter rebased

jumbi77 commented:
@tuxoko Can I may ask for another rebase and @tonyhutter for review? Get this done would be great! In any case much thanks.

tonyhutter commented on module/zcommon/zfeature_common.c:
```diff
@@ -754,6 +754,19 @@ zpool_feature_init(void)
+		};
+		zfeature_register(SPA_FEATURE_LONGNAME,
+		    "org.zfsonlinux:longname", "longname",
+		    "support filename upto 1024 bytes",
```
"up to"

tonyhutter commented on module/zfs/zap_micro.c:
```diff
@@ -149,43 +149,104 @@ zap_normalize(zap_t *zap, const char *name, char *namenorm, int normflags)
+		char normbuf[ZAP_MAXNAMELEN];
+		char *norm = normbuf;
 
-		if (zap_normalize(zn->zn_zap, matchname, norm,
-		    zn->zn_normflags) != 0)
-			return (B_FALSE);
+		/*
+		 * Cannot allocate this on-stack as it exceed the stack-limit of
+		 * 1024.
+		 */
+		if (namelen > ZAP_MAXNAMELEN)
+			norm = kmem_alloc(namelen, KM_SLEEP);
```
Can we get rid of normbuf[] all together, or is there there a benefit to using the stack when possible?

tonyhutter replied:
Just thinking out loud - I'm curious if it would be worth it to do:
```diff
- norm = kmem_alloc(namelen, KM_SLEEP);
+ norm = kmem_cache_alloc(zap_name_long, KM_SLEEP);
```
You would waste some memory, but allocations might be faster.  I dunno if zap_match() is a critical path or not though.

tuxoko replied:
> Can we get rid of normbuf[] all together, or is there there a benefit to using the stack when possible?

I'm ok with removing. I think it only matters if case insensitive or some form of utf8 normalization is inuse.

tonyhutter commented:
> Note that, to my knowledge native Linux filesystems don't support name longer than 255 bytes. So there might be programs not able to work with longname.

That seems to be the case.  I'm unable to write a >256B filename on Fedora 40 using this PR :
```
/tank$ touch $(printf 'a%.0s' {1..255})
/tank$ touch $(printf 'a%.0s' {1..256})
touch: cannot touch 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa': File name too long
```
So what would be the use case for 1024B filenames if it's not supported by Linux?  Does it work in FreeBSD?

Haravikk commented:
> So what would be the use case for 1024B filenames if it's not supported by Linux? Does it work in FreeBSD?

IIRC the limit for Linux is 255 characters, including multi-byte characters, but the current ZFS limit is 255 bytes, so multibyte characters will eat into that. For example if you're using two-byte characters your limit is only 127 characters (254 bytes).

There are a bunch of filesystems now that support either 255 characters, or have even less strict limits (so they'll support whatever the OS' limit(s) are).

It's also one of those classic chicken and egg problems – if ZFS keeps a limitation because it matched one on Linux, then Linux is less likely to increase that limit because a major filesystem doesn't support it etc.

IMO it's better for a filesystem to support any size that it doesn't have a good technical reason not to (due to overhead or whatever), that way it's not the part that's holding something back. Though it makes sense to have it tunable for those that need to guarantee portability between more and less limited OSes.

tuxoko commented:
> > Note that, to my knowledge native Linux filesystems don't support name longer than 255 bytes. So there might be programs not able to work with longname.
> 
> That seems to be the case. I'm unable to write a >256B filename on Fedora 40 using this PR :
> 
> ```
> /tank$ touch $(printf 'a%.0s' {1..255})
> /tank$ touch $(printf 'a%.0s' {1..256})
> touch: cannot touch 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa': File name too long
> ```
> 
> So what would be the use case for 1024B filenames if it's not supported by Linux? Does it work in FreeBSD?

@tonyhutter 
Did you enable both the pool feature and turn on the property on the dataset?

The kernel VFS layer itself doesn't impose a 255 limit on filename as long as the "path" length fits in 4k page.
As you can see the unit tests are passed in the built bot, there shouldn't be any problem creating longname file with common shell command. Any program that rely only on syscall to check file name length should work fine.
We have no problem running this with samba.
The only notable issue if nfs server may need to call exportfs_get_name, and it has a hard coded buffer size of 256.

```
$ touch $(printf 'a%.0s' {1..255})
$ touch $(printf 'a%.0s' {1..256})
$ ls
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa
```

tonyhutter commented:
@tuxoko thanks, you were right, I wasn't enabling the `longname` dataset feature.  Its working fine for me natively in Linux.  :+1:

tonyhutter commented:
I'm pretty happy with this - if you want to rebase it I can approve it.

behlendorf commented on tests/zfs-tests/tests/functional/longname/longname_001_pos.ksh:
```diff
@@ -0,0 +1,139 @@
+INODE_NO_DIR=$(stat -c %i $WORKDIR/$LONGDNAME)
+INODE_NO_LINK=$(stat -c %i $WORKDIR/$LONGLNAME)
+INODE_NO_PIPE=$(stat -c %i $WORKDIR/$LONGPNAME)
+INODE_NO_CHAR=$(stat -c %i $WORKDIR/$LONGCNAME)
```
You can use `get_objnum` which works on Linux and FreeBSD.  Although it looks like the values fetched here are never used.

tuxoko replied:
Yeah, these are leftovers. I'll remove them.

behlendorf commented on tests/runfiles/linux.run:
```diff
@@ -138,6 +138,10 @@ pre =
 
+[tests/functional/longname:Linux]
+tests = ['longname_001_pos', 'longname_002_pos', 'longname_003_pos']
+tags = ['functional', 'longname']
```
Is there a reason we can't enable these on FreeBSD?

tuxoko replied:
I don't know if it would work on FreeBSD or not. I guess I can let buildbot to test it.

behlendorf replied:
Yes, let's see what happened on FreeBSD.

nit: I see there's also a [cstyle warning](https://github.com/openzfs/zfs/actions/runs/10190884366/job/28191292745?pr=15921) to sort out.

behlendorf reviewed (commented):
@tuxoko sorry about the slow review on this.  Just a couple of minor nits then this looks good to me.  If you can get this rebased I can pull it in.

robn commented:
I had a few of drive-by questions/thoughts. Note I don't consider these blockers; just thinking about the problem space.

Is it worth storing the max filename length somewhere on the dir ZAP (extended attribute? somewhere on the bonus?) so that we can increase it further in the future without needing another feature flag etc?

I assume we're happy leaving leaving interpretation of the bytes in the filename as out of scope (ie, normal `utf8only`, `normalization`, `casesensitivity`, etc)?

Is there any value in a kind of "compatibility" option, so that older implementations can still use these dirs? I personally think not; it would be a lot of complexity for perhaps very little advantage. And its tricky to do well; at least, I remember FAT32 long filenames being surprisingly difficult to get right :)

amotin commented on module/os/linux/zfs/zpl_inode.c:
```diff
@@ -151,6 +181,10 @@ zpl_create(struct inode *dir, struct dentry *dentry, umode_t mode, bool flag)
 	zidmap_t *user_ns = kcred->user_ns;
 #endif
 
+	if (is_nametoolong(dentry)) {
+		return (-ENAMETOOLONG);
+	}
```
Cosmetics, but I prefer to avoid extra braces for a single line.

amotin commented on module/os/linux/zfs/zpl_xattr.c:
```diff
@@ -153,26 +153,28 @@ static int
 {
 	zap_cursor_t zc;
-	zap_attribute_t	zap;
+	zap_attribute_t	*zap = zap_attribute_alloc();
```
Again to the question of extended attribute names length limit.

tuxoko replied:
xattr name length is limit by kernel to 255

amotin replied:
OK if so, I just haven't found the check on my quick search for it.

tuxoko replied:
https://elixir.bootlin.com/linux/v6.11/source/include/uapi/linux/limits.h#L15

amotin commented on module/zcommon/zfeature_common.c:
```diff
@@ -760,6 +760,19 @@ zpool_feature_init(void)
+	{
+		static const spa_feature_t longname_deps[] = {
+			SPA_FEATURE_EXTENSIBLE_DATASET,
+			SPA_FEATURE_ENABLED_TXG,
```
Do we really care at what TXG long names were actually enabled?  What do we do with this knowledge?

tuxoko replied:
I'm not sure if there's any reason when this was added. Maybe for audit purpose. Anyway I'll remove it.

amotin replied:
Thanks. I am not sure what exactly does the extensible dataset, so I just hope it is there for reason and not just copy-paste.

amotin commented on module/zcommon/zfs_prop.c:
```diff
@@ -772,6 +772,10 @@ zfs_prop_init(void)
 	    ZFS_TYPE_VOLUME, "<date>", "SNAPSHOTS_CHANGED", B_FALSE, B_TRUE,
 	    B_TRUE, NULL, sfeatures);
 
+	zprop_register_index(ZFS_PROP_LONGNAME, "longname", 0, PROP_INHERIT,
+	    ZFS_TYPE_FILESYSTEM, "on | off", "LONGNAME", boolean_table,
+	    sfeatures);
```
I am somewhat split about this property and its disabled default.  I may speculate that some users may not want this functionality, while same time once it become routine I don't think I want to manually enable it on each new pool created.

tuxoko replied:
The reason it's not enabled is because it may not be suitable for all workload. For example NFS, or when you copy files between native linux fs.

lundman replied:
> I am somewhat split about this property and its disabled default. 
I am hoping to have a couple of days after it lands to see if I can make it work on my platforms, before it is made default ...

behlendorf commented:
@tuxoko the updated ABI files look like they contain a bunch on unrelated changes (and they they pass).  If you remove the ABI file changes, then force update the PR you should be able to download new ABI files from the "checkstyle" builder summary section with the minimal changes required.  e.g. generated with the exact same version of libabigail.

amotin commented on module/zfs/zfs_ioctl.c:
```diff
@@ -2594,6 +2594,44 @@ zfs_prop_set_special(const char *dsname, zprop_source_t source,
+		zfsvfs_t *zfsvfs;
+		dsl_dataset_t *ds;
+
+		/*
+		 * Ignore the checks if the property is being applied as part of
+		 * 'zfs receive'. Because, we already check if the local pool
+		 * has SPA_FEATURE_LONGNAME enabled in dmu_recv_begin_check().
+		 */
+		if (source == ZPROP_SRC_RECEIVED) {
+			cmn_err(CE_NOTE, "Skipping ZFS_PROP_LONGNAME checks "
+			    "for dsname=%s\n", dsname);
+			err = -1;
+			break;
+		}
```
Do we really need this special case?  Received stream may be inconsistent/malicious.

tuxoko replied:
Right, I'll remove it.

tuxoko replied:
Actually, I had to add this back because when recving a new dataset, zfsvfs_hold would fail.
Maybe we shouldn't call zfsvfs_hold here, but I'll just restore to the original code.
I don't think there would be problem with malformed stream anyway because as the comment says we already checked longname feature, so there's no problem with enable the property on dataset.

tuxoko commented on module/os/linux/zfs/zpl_inode.c:
```diff
@@ -46,9 +46,29 @@ zpl_lookup(struct inode *dir, struct dentry *dentry, unsigned int flags)
+		 */
+		if (!dsl_dataset_feature_is_active(ds, SPA_FEATURE_LONGNAME) ||
+		    (flags & (LOOKUP_CREATE | LOOKUP_RENAME_TARGET)))
+			return (ERR_PTR(-ENAMETOOLONG));
```
I think if it's not active then we can just return early.

amotin replied:
I don't think this is frequent enough (outside some special synthetic) case to optimize for it, and it just complicated the logic.

amotin commented on module/os/freebsd/zfs/zfs_dir.c:
```diff
@@ -368,18 +370,19 @@ static int
 	int skipped = 0;
 	int error;
 
+	zap = zap_attribute_alloc();
```
I see this is used only for extended attributes directory, so IF they are really limited to 255 chars, this may be fine.  Though can't say that I like IF's like that.

tuxoko replied:
fixed

tuxoko replied:
Yeah, I revert it back. I forgot it's only for xattr like you said. So it's not needed.

tuxoko commented on module/os/linux/zfs/zfs_dir.c:
```diff
@@ -589,7 +590,7 @@ static int
 {
 	zap_cursor_t	zc;
-	zap_attribute_t	zap;
+	zap_attribute_t	*zap = zap_attribute_alloc();
```
fixed

tuxoko replied:
Yeah, I revert it back. I forgot it's only for xattr like you said. So it's not needed.

behlendorf commented:
Updated ABI files available: https://github.com/openzfs/zfs/actions/runs/11022167809?pr=15921

tuxoko commented:
Updated with the abi files and freebsd compile fix. Hopefull we can get a freebsd run now.

A couple of notable new changes,
I added check in fzap_checkname to make sure we don't accidentally use longname for non directory zap.
Also remove readonly compat for SPA_FEATURE_LONGNAME, as it might cause problem.

tuxoko commented on module/zfs/zap.c:
```diff
@@ -832,7 +832,11 @@
 {
-	if (zn->zn_key_orig_numints * zn->zn_key_intlen > ZAP_MAXNAMELEN)
+	uint32_t maxnamelen = zn->zn_normbuf_len;
+	uint64_t len = zn->zn_key_orig_numints * zn->zn_key_intlen;
```
fxied

amotin commented on module/zfs/zap_micro.c:
```diff
@@ -149,43 +149,96 @@ zap_normalize(zap_t *zap, const char *name, char *namenorm, int normflags)
+		size_t namelen = zn->zn_normbuf_len;
+		char *norm;
 
+		norm = kmem_alloc(namelen, KM_SLEEP);
```
It seems `zap_match()` is called in a loop over potentially many other entries. Allocation for might be expensive.  Sure using stack previously was a cheating, but this might need some more thinking.

tuxoko replied:
I can go back to the original one.

tuxoko commented:
Update, fixed compile error and a typo in test.
Also, it turns out freebsd has 255 limit in vfs layer.
https://github.com/freebsd/freebsd-src/blob/01eb635d12953e24ee5fae69692c28e4aab4f0f6/sys/kern/vfs_lookup.c#L1132
So this won't actually work on freebsd, so I moved the unit tests back to linux only.

behlendorf commented on module/zcommon/zfeature_common.c:
```diff
@@ -760,6 +760,18 @@ zpool_feature_init(void)
+		    "org.zfsonlinux:longname", "longname",
+		    "support filename up to 1024 bytes",
+		    ZFEATURE_FLAG_PER_DATASET, ZFEATURE_TYPE_BOOLEAN,
+		    longname_deps, sfeatures);
```
Should be added to `man/man7/zpool-features.7` in alphabetical order.

tuxoko commented on module/os/freebsd/zfs/zfs_znode_os.c:
```diff
@@ -1814,7 +1814,7 @@ zfs_znode_parent_and_name(znode_t *zp, znode_t **dzpp, char *buf)
 
 	err = zap_value_search(zfsvfs->z_os, parent, zp->z_id,
-	    ZFS_DIRENT_OBJ(-1ULL), buf);
+	    ZFS_DIRENT_OBJ(-1ULL), buf, MAXNAMELEN);
```
Yeah, it was hardcoded as well in zap_value_search before. Though in this case we can make the caller to pass buf length.

tuxoko commented on module/zfs/zap.c:
```diff
@@ -832,7 +832,11 @@ zap_put_leaf_maybe_grow_ptrtbl(zap_name_t *zn, zap_leaf_t *l,
+	uint64_t len = zn->zn_key_orig_numints * zn->zn_key_intlen;
+	if (len > maxnamelen ||
+	    (zn->zn_zap->zap_dnode->dn_type != DMU_OT_DIRECTORY_CONTENTS &&
+	    len > ZAP_MAXNAMELEN))
```
Yeah, it's not elegant, but it's readily available.

behlendorf commented:
@tuxoko looks good.  The CI failures here all look to be unrelated.  I can handle resolving the trivial conflict in the merge.

tschoening81 commented:
Maybe I missed it, but could someone please explain how the implementation actually works currently? The PR is just too large and in-depth for me to answer the questions I have. Thanks!

# 8.3-approach

In some other related issues a suggestion for example was to follow the Windows 8.3 naming scheme for backwards compatibility reason, storing the real long name additionally in some xattr. But that would have influence on tools looking at filenames, which I don't see documented in the feature description currently.

So, there's no special backwards compatible workaround and a long filename is simply stored in a larger buffer, but still as the one and only filename? Tools always see one and the same name only and need to support a possibly longer name?

# Performance impact

OTOH, if longer names are stored sperately with/without an approach like 8.3 for the original name, that might influence performance. Such an approach might be worth it mentioning a possible performance impact in the docs as well in my opinion. As I don't see it in the docs, that is not the implemented approach or do you just disagree the need of documenting those things?

What I have in mind is e.g. encryption, which made RSYNC based filesystem iteration a lot slower in my setup as well. There are some discussions about other people having similar observations and that it might just be like that because of how encryption works. But that slowdown was somewhat surprising to me and I would have liked a mention of it in the docs beforehand.

# Read-only

There's the following comment:

> Also remove readonly compat for SPA_FEATURE_LONGNAME, as it might cause problem.

https://github.com/openzfs/zfs/pull/15921#issuecomment-2380343195

How is that related to the current docs?

> This property can be disabled even if it contains longname files. In such case, new file cannot be created with longname but existing longname files can still be looked up.

https://openzfs.github.io/openzfs-docs/man/master/7/zpool-features.7.html#longname

beedell-roke commented:
> This patch adds the ability for zfs to support file/dir name up to 1023
bytes. This number is chosen so we can support up to 255 4-byte
characters.

Does no way for the FS to support arbitrary lengths exist, > 1023 characters?
